File: TEO/Mav.UI/Mav.Ui/src/app/services/i-question.service.ts
import { Observable } from 'rxjs/Observable';
import { GenericResult } from '../models/generic.result.model';
import { QuestionViewModel } from '../models/question.view.model';
import { AddQuestionViewModel } from '../models/admin/add-question.view.model';
import { UpdateQuestionViewModel } from '../models/admin/update-question.view.model';

export interface IQuestionService {

    getQuestionById<T>(id: number): Observable<T>;

    getQuestionList(sectionId: number): Observable<QuestionViewModel[]>;

    createQuestion(newItem: AddQuestionViewModel): Observable<GenericResult<number>>;

    updateQuestion(item: UpdateQuestionViewModel): Observable<GenericResult<number>>;

    deleteQuestion(id: number): Observable<GenericResult<number>>;

}

File: TEO/Mav.Services/Abstract/IQuestionService.cs
using Mav.Common.Models;
using Mav.Models.ViewModels;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Mav.Services.Abstract
{
    public interface IQuestionService
    {
        Task<T> GetQuestionById<T>(int id);

        Task<ExamQuestionViewModel> GetExamQuestionById(int id);

        Task<IEnumerable<T>> GetQuestionList<T>(int sectionId);

        Task<GenericResult<T>> CreateQuestion<T>(AddQuestionViewModel m);

        Task<GenericResult<T>> UpdateQuestion<T>(UpdateQuestionViewModel m);

        Task<GenericResult<T>> DeleteQuestion<T>(int id);

        Task<IEnumerable<T>> GetQuestionBySectionId<T>(int sectionId);
    }
}
